server/asset/eth/coiner.go

```diff
@@ -0,0 +1,284 @@
+	// hash, value should be checked by using the contract once it has
+	// confirmations.
+	value, err := ToGwei(tx.Value())
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert value: %v", err)
+	}
+
+	// For redemptions, the transaction should move no value.
+	if sct == sctRedeem && value != 0 {
+		return nil, fmt.Errorf("expected swapCoin value of zero for redeem but got: %d", value)
+	}
+
+	return &swapCoin{
+		backend:      backend,
+		contractAddr: *contractAddr,
+		secret:       secret,
+		secretHash:   secretHash,
+		value:        value,
+		gasPrice:     gasPrice,
+		txid:         hex.EncodeToString(txid[:]),
+		counterParty: *counterParty,
+		locktime:     locktime,
+		sct:          sct,
+	}, nil
+}
+
+// Confirmations returns the number of confirmations for a Coin's
+// transaction. Because a Coin can become invalid after once being
+// considered valid, this condition should be checked for during
+// confirmation counting and an error returned if this Coin is no longer
+// ready to spend. An unmined transaction should have zero confirmations. A
+// transaction in the current best block should have one confirmation. A
+// negative number can be returned if error is not nil.
+//
+// In the case of ethereum it is extra important to check confirmations before
+// confirming a swap. Even though we check the initial transaction's data, if
+// that transaction were in mempool at the time, it could be swapped out with
+// any other values if a user sent another transaction with a higher gas fee
+// and the same account and nonce, effectivly voiding the transaction we
+// expected to be mined.
+func (c *swapCoin) Confirmations(_ context.Context) (int64, error) {
+	swap, err := c.backend.node.swap(c.backend.rpcCtx, c.secretHash)
+	if err != nil {
+		return -1, err
+	}
+
+	assertMempool := func() error {
+		_, mempool, err := c.backend.node.transaction(c.backend.rpcCtx, common.HexToHash(c.txid))
+		if err != nil {
+			return fmt.Errorf("unable to fetch transaction: %v", err)
+		}
+		if !mempool {
+			return fmt.Errorf("critical problem with swap coin: tx not in mempool but swap state is %s", SwapState(swap.State))
+		}
+		return nil
+	}
+
+	switch c.sct {
+	case sctRedeem:
+		// There should be no need to check the counter party, or value
+		// as a swap with a specific secret hash that has been redeemed
+		// wouldn't have been redeemed without ensuring the initiator
+		// is the expected address and value was also as expected. Also
+		// not validating the locktime, as the swap is redeemed and
+		// locktime no longer relevant.
+		ss := SwapState(swap.State)
+		if ss == SSRedeemed {
+			// While not completely accurate, we know that if the
+			// swap is redeemed the redemption has at least one
+			// confirmation.
+			return 1, nil
+		}
+		// If swap is in the Initiated state, the transaction must be
+		// unmined.
+		if ss == SSInitiated {
+			if err := assertMempool(); err != nil {
```

Comment:
Can you clarify this for me?  Why must it be in mempool?
If this is a redeem coin/txn, and the contract claims that the state is in initiated still, I think that makes sense.  However, this looks like a race to me -- get swap state, if "initiated" get redeem transaction mempool status.  If in initiated state and this is a redeem coin confirmations check, would it hurt to just report 0 without asserting mempool?

Reply:
Just assuming the tx is still in mempool if we don't see the contract changing seems like the cleanest solution for now. Removed assertion.